modules/gdscript/README.md

```diff
@@ -0,0 +1,92 @@
+2. A tightly designed language. Features are added because they are _needed_, and not because they can be added or are interesting to develop.
+3. Primarily an interpreted scripting language. It is meant to be easy to use and develop gameplay in. It is not meant for CPU-intensive algorithms or data processing, and is not optimized for it. For that, [C#](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_basics.html) or [GDExtension](https://docs.godotengine.org/en/stable/tutorials/scripting/gdextension/what_is_gdextension.html) may be used.
+
+
+## Compilation
+Scripts can be at different stages of compilation, but the process generally consists of a linear series of steps: tokenizing, parsing, analyzing, and finally compiling.
+
+The main class of the GDScript module is the [`GDScript`](gdscript.h) class, which represents a class defined in GDScript. Each `.gd` file is called a _class file_ because it implicitly defines a class in GDScript, and thus results in an associated `GDScript` object. However, GDScript classes may define [_inner classes_](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#inner-classes), and those are also represented by further `GDScript` objects, even though they are not in files of their own.
+
+The `GDScript` class contains all the information related to the corresponding GDScript class: its name and path, its members like variables, functions, symbols, signals, implicit methods like initializers, etc. This is the main class that the compilation step deals with.
+
+A secondary class is `GDScriptInstance`, defined in the same file, containing _runtime_ information for an instance of a `GDScript`, and is more related to the execution of a script by the virtual machine.
+
+### Loading source code
+
+This mostly happens by calling `GDScript::load_source_code()` on a `GDScript` object. Parsing only requires a `String`, so it is entirely possible to parse a script without a `GDScript` object!
+
+### Tokenizing (see [`gdscript_tokenizer.h/cpp`](gdscript_tokenizer.h))
+
+Tokenizing is the process of converting the source code `String` into a sequence of tokens, which represent language constructs, identifiers, literals, etc. It happens as the first step of the parsing process, and so isn't really found independently of that in the codebase, save very rare exceptions.
+
+### Parsing (see [`gdscript_parser.h/cpp`](gdscript_parser.h))
+
+The parser takes a sequence of tokens and builds the abstract syntax tree (AST) of the GDScript program. This is what finds any syntax errors in a GDScript program. The AST is used in the analyzing and compilation steps, and the source code `String` and sequence of tokens are discarded. 
+
+The parser class also defines all the possible nodes of the syntax tree, e.g. `GDScriptParser::ClassNode` for a node representing a GDScript class. It further defines the typechecking datastructures like `GDScriptParser::Datatype`.
```

Comment:
It may be worth mentioning the parser limitations that we [deliberately added](https://godotengine.org/article/gdscript-progress-report-writing-new-parser/#less-lookahead). The parser makes **one** pass: unlike the analyzer, some nodes have not yet been parsed. At any given time, only 2 tokens are available to the parser: the previous and the current. Understanding these facts has helped me in several PRs when I was trying to do something that was impossible.

Reply:
Yes!!!! I totally forgot that that was a super intentional design decision that is _absolutely_ worth mentioning here!